### ISSUE TITLE
[RFC] Accessibility architecture proposal

### ISSUE THREAD

**Issue by smykhailov** [stardust-ui/react-old#119]
## What is accessibility

* sufficient contrast in standard themes
* high contrast theme
* zoom
* keyboard navigation / focus handling / focus highlighting
* screen reader support for virtual cursor and narration

## Goal

* each component must be accessible
* each component should have correct color contrast and high contrast theme defined
* accessibility behavior should be separated from the component
* accessibility behaviors should be extendable and reusable
* component should have its own default accessibility behavior, which can be overridden from parent component, if needed
* accessibility behavior should be easy to use

## Stardust and accessibility

* Stardust is a flexible UI library that allows composition of components. Accessibility patterns are much more restrictive. We need a way of applying the restrictive patterns to the flexible UI components
* Stardust should provide accessibility by default, but the developer using Stardust should be able to modify/override the behavior to fit his specific needs
* All components need to be semantically correct
* RTL has impact on keyboard navigation as it switches right/left arrow key functionality

## Statement

**Component** uses **behavior** to apply event handlers and aria properties to **HTML elements**.

The **behavior** of a component is initially determined by the component type and properties.
User of the component (developer) can **override** this default behavior by providing a different behavior type in the properties of the component.

There are two parts of accessibility, which should be encapsulated in the behaviors:

* `aria-*` attributes - specifies the set of attributes necessary for correct narrating by screen readers.
* keyboard navigation - the way how to use application without mouse, whenever user navigates with keyboard he/she should be able to see where the focus is. The focus can be trapped in some areas (e.g. modal dialogs).

Elements respond to HTML keyboard events by calling corresponding methods specified by the behavior.

Behavior can execute an action (for example move focus to the next child component), and/or create an **intent** and propagate this intent to the parent component behavior.

Parent component behavior can react to this intent by executing an action and/or propagating the intent to its parent component behavior.

Intent propagation is explicit (as opposed to HTML event propagation). Behavior needs to explicitly send the intent to the parent behavior.

For HTML events, once the element responds to an event by calling behavior method, the event propagation should be stopped.

## Accessibility in HTML

* semantically correct HTML, to support Screen Reader's virtual cursor navigation
* proper use of `aria-*` attributes and tabindex attributes, to support Screen Readers narration
* keyboard handling implemented in JavaScript, for the dynamic content, such as popup menus, menus, lists, dialogs, etc.

## Themes and color contrast

* each component should have at least two themes: standard and high contrast, to support this we use CSS -in-JS technique
* each component's color palette should be verified for sufficient contrast

## Behaviors

The idea behind **behaviors** is to decouple accessibility logic from the component and make it reusable, easy overridable and extendable.

Base component should expose the strongly typed accessibility behavior property, which has to be assigned by each component.

```TypeScript
export class UIComponent<P, S> extends React.Component<P, S> { 
  …
  public accBehavior: IAccessibilityBehavior<P, S>
  …
}
``` 

The behavior should take properties and state as a parameters, to generate correct `aria-*` attributes and set proper tabindex. 

```TypeScript
export interface IAccessibilityBehavior<P, S> {
  readonly name: string
  generateAriaAttributes(props: P, state: S): object
  // the following line might not be needed after component state management is introduced
  changeState(newState: ComponentState): void
}
```

Stardust framework should provide a set of predefined behaviors, which can be easily extended if needed. Very simplified version, just to show the idea behind it, how it might look like:

```TypeScript
export class ButtonBehavior implements IAccessibilityBehavior<{}, {}> {
  private attributes = {
    'ms-acc-behavior': this.name,
    role: 'button',
    'aria-hidden': false,
	// child: {
	//   'ms-acc-behavior': this.name,
	//   role: 'menu-item',
	//   'aria-hidden': false,
	//   'tab-index': 0,
	// },
  }

  public get name(): string {
    return 'button'
  }

  public generateAriaAttributes(): object {
    return this.attributes
  }

  public changeState(newState: ComponentState): void {
    if (newState === ComponentState.disabled) {
      this.attributes['aria-disabled'] = true
    } else if (newState === ComponentState.enabled) {
      delete this.attributes['aria-disabled']
    }
  }
}
``` 

The component should instantiate its behavior:

```TypeScript
constructor(p, s) {
  super(p, s)
  this.accBehavior = new ButtonBehavior()
}
``` 

and render component with generated attributes:

```TypeScript
renderComponent({ ElementType, classes, rest }) {
  const { children, content } = this.props

  return (
    <ElementType
      {...this.accBehavior.generateAriaAttributes(this.props, this.state)}
      {...rest}
     className={classes.root}
   >
     {childrenExist(children) ? children : content}
    </ElementType>
  )
}
```

`{…rest}` goes after behavior generated attributes, to give more flexibility and allow to override anything generated by the behavior if developers need it.

Whenever component changes its state, the new `aria-*` attributes will be provided back to the component by behavior. 

// TODO: extend the proposal for keyboard navigation and focus management

## Supported client/screen reader combinations

* Win/Desktop - JAWS
* Win/Chrome - JAWS
* Win/Edge - Narrator
* Win/Firefox - NVDA
* Mac/Desktop - VoiceOver

## Glossary

**Component** - Stardust (react) component
**Behavior** - Javascript function,object or class that defines the behavior of a component
**Element** - HTML element rendered by the component
**Event** - HTML event (for example keydown)
**Intent** - operation to be performed. Behaviors understand intents.

**Comment by jurokapsiar** [comment]
Please look at the **Examples** section in https://github.com/stardust-ui/react/pull/105/files which was left out from this particular description but provides a concrete example of how the parts tie together

**Comment by jurokapsiar** [comment]
closing as this will be moved to the new repo
